src/devices/Gpio/Drivers/VirtualGpioController.cs

```diff
@@ -0,0 +1,279 @@
+        private readonly ConcurrentDictionary<int, PinEvent> _callbackEvents = new ConcurrentDictionary<int, PinEvent>();
+
+        private record PinEvent(PinEventTypes PinEventTypes, PinChangeEventHandler? Callbacks);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VirtualGpioController"/> class.
+        /// </summary>
+        public VirtualGpioController()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VirtualGpioController"/> class.
+        /// </summary>
+        /// <param name="numberingScheme">The numbering scheme used to represent pins provided by the controller.</param>
+        public VirtualGpioController(PinNumberingScheme numberingScheme)
+        {
+            // Nothing on purpose, as only logical is suported
+            if (numberingScheme != PinNumberingScheme.Logical)
+            {
+                throw new ArgumentException("Only PinNumberingScheme Logical is supported.");
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VirtualGpioController"/> class.
+        /// </summary>
+        public VirtualGpioController(Dictionary<int, GpioPin> pins)
+        {
+            foreach (var pin in pins)
+            {
+                _pins.TryAdd(pin.Key, pin.Value);
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VirtualGpioController"/> class.
+        /// </summary>
+        public VirtualGpioController(IEnumerable<GpioPin> pins)
+        {
+            int inc = 0;
+            foreach (var pin in pins)
+            {
+                _pins.TryAdd(inc++, pin);
+            }
+        }
+
+        /// <summary>
+        /// Adds a new pin number with an associated <see cref="GpioPin"/>.
+        /// </summary>
+        /// <param name="pinNumber">The pin number.</param>
+        /// <param name="gpioPin">The <see cref="GpioPin"/> to add.</param>
+        public void Add(int pinNumber, GpioPin gpioPin) => _pins.TryAdd(pinNumber, gpioPin);
+
+        /// <summary>
+        /// Adds a new <see cref="GpioPin"/>. The number is done automatically by adding after the last element.
+        /// </summary>
+        /// <param name="gpioPin">The <see cref="GpioPin"/> to add.</param>
+        public void Add(GpioPin gpioPin)
+        {
+            // Find the last number of the pin and assign the new number to it
+            int newPin = _pins.Count > 0 ? _pins.Keys.Max() + 1 : 0;
+            _pins.TryAdd(newPin, gpioPin);
+        }
+
+        /// <inheritdoc/>
+        public override int PinCount => _pins.Count;
+
+        /// <summary>
+        /// Closes an open pin.
+        /// If allowed by the driver, the state of the pin is not changed.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        public override void ClosePin(int pinNumber)
```

Comment:
Implementation or description is missleading, as you don't actually close the underlying pin.